add subscriber for JointTrajectory as /command

k-okada opened the pull request:
/command is general API[1] that is widly used together with JointTrajectoryAction, I need to use this IF for calibration program[2]

[1] http://wiki.ros.org/joint_trajectory_controller?distro=indigo
[2] https://github.com/ros-perception/calibration/pull/31

k-okada committed:
add subscriber for JointTrajectory as /command

Commit:
Merge pull request #71 from RethinkRobotics/release-1.2.0

Release 1.2.0

k-okada commented:
resolved conflicts

IanTheEngineer commented on src/joint_trajectory_action/joint_trajectory_action.py:
```diff
@@ -68,6 +69,7 @@ def __init__(self, limb, reconfig_server, rate=100.0,
             FollowJointTrajectoryAction,
             execute_cb=self._on_trajectory_action,
             auto_start=False)
+        self._command = rospy.Subscriber('robot/limb/' + limb + '/command', JointTrajectory, self._on_joint_trajectory)
```
I would advocate for the following namespacing for clarity:

```
/robot/limb/right/joint_trajectory/command
/robot/limb/right/joint_trajectory/state
/robot/limb/right/joint_trajectory/query_state
```

k-okada replied:
thanks for comment, fixed,

IanTheEngineer commented on src/joint_trajectory_action/joint_trajectory_action.py:
```diff
@@ -488,3 +490,116 @@ def check_goal_state():
+                self._pid[jnt].initialize()
+
+    def _on_joint_trajectory(self, trajectory):
+        joint_names = trajectory.joint_names
```
As this trajectory callback can execute at the same time as the Action callback, I think it would make sense to have a mutex lock on both to prevent simultaneous joint commands from being sent into the robot.

IanTheEngineer commented:
@k-okada I'm very sorry your pull request has languished for nearly two years :/ . I think this topic interface is a valuable addition to the joint trajectory action server, and want to incorporate it if possible. Since the `master` branch is considered "live", can you target your pull requests at the `development` branch?

I am off-the-grid the next week on vacation, but I would like to address your pull requests the first week in August when I come back.

k-okada committed:
add mutex to avoid race condition

k-okada commented:
ok, i have rebased on `development` branch and add mutex lock code. I added mutex lock within `on_trajectory_action` method (https://github.com/RethinkRobotics/baxter_interface/pull/37/commits/5b954f97527be52ae58b98db518eac9cc40f19fc#diff-393fafc5e38a00ca79cd9d500e6ab03cR350) at https://github.com/RethinkRobotics/baxter_interface/pull/37/commits/5b954f97527be52ae58b98db518eac9cc40f19fc#diff-393fafc5e38a00ca79cd9d500e6ab03cR421, based on `# Acquire Mutex` comment
(the lock code is removed at https://github.com/RethinkRobotics/baxter_interface/pull/51/files), and within `on_trajectory_command` method at https://github.com/RethinkRobotics/baxter_interface/pull/37/commits/5b954f97527be52ae58b98db518eac9cc40f19fc#diff-393fafc5e38a00ca79cd9d500e6ab03cR582

But I do not think this will prevent simultaneous joint commands execution, we have to lock more wider range, i.e. entire on_trajectory_action and on_trajectory_command, and also I think we should support "override / trajectory replacement" feature as described in http://wiki.ros.org/robot_mechanism_controllers/JointTrajectoryActionController#Trajectory_Replacement, I'll work on this once all my PR is merged.

Have a nice vacation!
